# <font color=blue> Problem_02 : 
### <font color=red>Experiment with Linear Regression

### Explore Dataset

In [41]:
import numpy as np
from scipy.optimize import minimize
from scipy.io import loadmat
from numpy.linalg import det, inv
from math import sqrt, pi
import scipy.io
import matplotlib.pyplot as plt
import pickle
import sys

In [35]:
X,y,Xtest,ytest = pickle.load(open('diabetes.pickle','rb'),encoding = 'latin1')

# X --> Xtrain ( 242 by 64) --> Input data matrix
# y --> Ytrain ( 242 by 1)  --> Target vector

# Xtest --> ( 200 by 64 ) --> Testing
# ytest --> ( 200 by 1 )  --> Testing

# So, X and y will be used for training

In [36]:
print("Shpae of Xtrain : ",X.shape)
print("Shpae of ytrain : ",y.shape)
print("---------------------------")
print("Shpae of Xtest : ",Xtest.shape)
print("Shpae of ytest : ",ytest.shape)

Shpae of Xtrain :  (242, 64)
Shpae of ytrain :  (242, 1)
---------------------------
Shpae of Xtest :  (200, 64)
Shpae of ytest :  (200, 1)


### Tasks :
- #### Implement <font color = blue >learnOLERegression ( )
- #### Implement <font color = blue >testOLERegression ( ) 

In [37]:
def learnOLERegression(X,y):
    '''
    Inputs:                                                         
          X = N x d  (Input data matrix for training)
          y = N x 1  (Target vector for training)                                                            
    Output: 
          w = d x 1   (Learned weight vector)
    '''
    
    # The formula for learning w in OLE : w = Inverse((Xtranspose * X)) * Xtranspose * y

    X_transpose   = np.transpose(X)
    X_X_transpose = np.dot(X_transpose,X)
    Inverse_X_X_transpose = np.linalg.inv(X_X_transpose) 
    
    w = np.dot(np.dot(Inverse_X_X_transpose,X_transpose),y)
    
    return w

In [38]:
def testOLERegression(w,Xtest,ytest):
    '''
    Inputs:
        w = d x 1
        Xtest = N x d
        ytest = X x 1
    Output:
        mse
    '''
    y_predict = np.dot(Xtest,w)
    
    diff = (ytest - y_predict)    
    diff_transpose = np.transpose(diff)
    N = 1 /len(Xtest)
    
    mse = np.dot( np.dot(N,diff_transpose), diff )
    
    return mse

### Main Code

In [39]:
# add intercept 
X_i     = np.concatenate((np.ones((X.shape[0],1)), X), axis=1)
Xtest_i = np.concatenate((np.ones((Xtest.shape[0],1)), Xtest), axis=1)

# Learn weight vector w from training data

w   = learnOLERegression(X,y)     # without intercept
w_i = learnOLERegression(X_i,y)   # with intercept

print("Test Data")
print("----------")

mle   = testOLERegression(w,Xtest,ytest)
mle_i = testOLERegression(w_i,Xtest_i,ytest)

print('MSE without intercept '+str(mle))
print('MSE with intercept '+str(mle_i))

print("\n")
 
print("Train Data")
print("----------")

mle   = testOLERegression(w,X,y)
mle_i = testOLERegression(w_i,X_i,y)

print('MSE without intercept '+str(mle))
print('MSE with intercept '+str(mle_i))


Test Data
----------
MSE without intercept [[106775.36155592]]
MSE with intercept [[3707.84018177]]


Train Data
----------
MSE without intercept [[19099.44684457]]
MSE with intercept [[2187.16029493]]


# <font color=blue> Report

## <font color=green>Case 01 : Without  an intercept 

#### MSE for test data : <font color=blue>  106775.36155592

#### MSE for training data : <font color=blue>  19099.44684457

## <font color=green>Case 01 : With an intercept 

#### MSE for test data : <font color=blue>  3707.84018177

#### MSE for training data : <font color=blue>  2187.16029493

### <font color = red> MSE is lower for the second case (using intercept) both for training data and test data. So, for the given dataset it is better to use an intercept for learning the weight vector for regression. Also we can observe that when we are using the intercept then the change in MSE is higher for the test data.